subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/ScriptSourceProvider.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.groovy.scripts.internal;
+
+import org.gradle.api.internal.artifacts.dsl.dependencies.UriTextResourceLoader;
+import org.gradle.groovy.scripts.ScriptSource;
+import org.gradle.groovy.scripts.UriScriptSource;
+
+import java.net.URI;
+
+public class ScriptSourceProvider {
+    private final UriTextResourceLoader uriTextResourceLoader;
+
+    public ScriptSourceProvider(UriTextResourceLoader uriTextResourceLoader) {
+        this.uriTextResourceLoader = uriTextResourceLoader;
+    }
+
+    public ScriptSource from(URI scriptUri) {
+        if ("http".equalsIgnoreCase(scriptUri.getScheme()) || "https".equalsIgnoreCase(scriptUri.getScheme())) {
```

Comment:
What schemes do we actually support here? We support S3 and SFTP for repositories: any reason they won't work for `apply from`? I'm thinking that by using our resource infrastructure, we might have improved support.

Another thing that we might have improved as a by-product of this PR: I think we will now be using the configured proxy settings for Gradle when downloading script resources.

Some test coverage for these would be good.